File: extensions/cpa-manager-plus/apps/web/src/services/api/adaptiveRouting.ts
import { apiClient } from './client';

export type AdaptiveRoutingConfig = {
  'top-k': number;
  'ewma-alpha': number;
  'ttft-target-ms': number;
  weights: {
    priority: number;
    load: number;
    'success-rate': number;
    ttft: number;
  };
  'sticky-escape': {
    enabled: boolean;
    'min-samples': number;
    'error-rate-threshold': number;
    'ttft-threshold-ms': number;
    'active-request-threshold': number;
  };
};

export type AdaptiveRoutingScore = {
  auth_id: string;
  auth_index: string;
  provider: string;
  priority: number;
  active_requests: number;
  success_rate: number;
  error_rate: number;
  ttft_ms: number;
  outcome_samples: number;
  ttft_samples: number;
  score: number;
  rank: number;
  eligible: boolean;
  escape_sticky: boolean;
  last_sample_at?: string;
};

export type AdaptiveRoutingSnapshot = {
  enabled: boolean;
  config: AdaptiveRoutingConfig;
  items: AdaptiveRoutingScore[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  has_more: boolean;
};

export type AdaptiveRoutingScoresQuery = {
  provider?: string;
  model?: string;
  search?: string;
  eligible?: boolean;
  page?: number;
  pageSize?: number;
};

const asRecord = (value: unknown): Record<string, unknown> =>
  value && typeof value === 'object' && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : {};

const finiteInteger = (value: unknown, fallback: number): number => {
  const parsed = typeof value === 'number' ? value : Number(value);
  return Number.isFinite(parsed) ? Math.trunc(parsed) : fallback;
};

export const normalizeAdaptiveRoutingSnapshot = (value: unknown): AdaptiveRoutingSnapshot => {
  const source = asRecord(value);
  const items = Array.isArray(source.items) ? (source.items as AdaptiveRoutingScore[]) : [];
  const total = Math.max(0, finiteInteger(source.total, items.length));
  const page = Math.max(1, finiteInteger(source.page, 1));
  const pageSize = Math.max(0, finiteInteger(source.page_size ?? source.pageSize, items.length));
  const totalPages = Math.max(
    0,
    finiteInteger(
      source.total_pages ?? source.totalPages,
      pageSize > 0 && total > 0 ? Math.ceil(total / pageSize) : 0
    )
  );
  const hasMoreValue = source.has_more ?? source.hasMore;

  return {
    enabled: source.enabled === true,
    config: asRecord(source.config) as AdaptiveRoutingConfig,
    items,
    total,
    page,
    page_size: pageSize,
    total_pages: totalPages,
    has_more:
      typeof hasMoreValue === 'boolean'
        ? hasMoreValue
        : String(hasMoreValue ?? '').toLowerCase() === 'true' || page < totalPages,
  };
};

export const adaptiveRoutingApi = {
  getConfig: () => apiClient.get<AdaptiveRoutingConfig>('/routing/adaptive'),
  updateConfig: (config: AdaptiveRoutingConfig) =>
    apiClient.put('/routing/adaptive', config),
  async getScores(query: AdaptiveRoutingScoresQuery = {}): Promise<AdaptiveRoutingSnapshot> {
    const data = await apiClient.get('/routing/adaptive/scores', {
      params: {
        provider: query.provider?.trim() || undefined,
        model: query.model?.trim() || undefined,
        search: query.search?.trim() || undefined,
        eligible: query.eligible,
        page: query.page,
        page_size: query.pageSize,
      },
    });
    return normalizeAdaptiveRoutingSnapshot(data);
  },
};

File: internal/config/adaptive_routing.go
package config

// AdaptiveRoutingConfig configures runtime-aware credential selection.
type AdaptiveRoutingConfig struct {
	TopK         int                    `yaml:"top-k,omitempty" json:"top-k,omitempty"`
	EWMAAlpha    float64                `yaml:"ewma-alpha,omitempty" json:"ewma-alpha,omitempty"`
	TTFTTargetMS int64                  `yaml:"ttft-target-ms,omitempty" json:"ttft-target-ms,omitempty"`
	Weights      AdaptiveRoutingWeights `yaml:"weights,omitempty" json:"weights,omitempty"`
	StickyEscape AdaptiveStickyEscape   `yaml:"sticky-escape,omitempty" json:"sticky-escape,omitempty"`
}

// AdaptiveRoutingWeights controls each normalized score component.
type AdaptiveRoutingWeights struct {
	Priority    float64 `yaml:"priority,omitempty" json:"priority,omitempty"`
	Load        float64 `yaml:"load,omitempty" json:"load,omitempty"`
	SuccessRate float64 `yaml:"success-rate,omitempty" json:"success-rate,omitempty"`
	TTFT        float64 `yaml:"ttft,omitempty" json:"ttft,omitempty"`
}

// AdaptiveStickyEscape controls when session affinity abandons a slow or unhealthy credential.
type AdaptiveStickyEscape struct {
	Enabled                bool    `yaml:"enabled,omitempty" json:"enabled,omitempty"`
	MinSamples             int64   `yaml:"min-samples,omitempty" json:"min-samples,omitempty"`
	ErrorRateThreshold     float64 `yaml:"error-rate-threshold,omitempty" json:"error-rate-threshold,omitempty"`
	TTFTThresholdMS        int64   `yaml:"ttft-threshold-ms,omitempty" json:"ttft-threshold-ms,omitempty"`
	ActiveRequestThreshold int     `yaml:"active-request-threshold,omitempty" json:"active-request-threshold,omitempty"`
}
